tests/test_date.py

```diff
@@ -2,66 +2,135 @@
+
+    @parameterized.expand([
+        param(begin=datetime(2014, 6, 15), end=datetime(2014, 6, 25), expected_length=10)
+    ])
+    def test_date_range(self, begin, end, expected_length):
+        self.given_period_start(begin)
+        self.given_period_end(end)
+        self.when_date_range_generated()
+        self.then_range_length_is(expected_length)
+        self.then_all_dates_are_present_in_range()
+        self.then_range_is_in_ascending_order()
+
+    @parameterized.expand([
+        param(begin=datetime(2014, 4, 15),
+              end=datetime(2014, 6, 25),
+              expected_months=[(2014, 4), (2014, 5), (2014, 6)]),
+        param(begin=datetime(2014, 4, 25),
+              end=datetime(2014, 5, 5),
+              expected_months=[(2014, 4), (2014, 5)]),
+        param(begin=datetime(2014, 4, 5),
+              end=datetime(2014, 4, 25),
+              expected_months=[(2014, 4)]),
+        param(begin=datetime(2014, 4, 25),
+              end=datetime(2014, 6, 5),
+              expected_months=[(2014, 4), (2014, 5), (2014, 6)]),
+    ])
+    def test_one_date_for_each_month(self, begin, end, expected_months):
+        self.given_period_start(begin)
+        self.given_period_end(end)
+        self.given_period_size(months=1)
+        self.when_month_range_generated()
+        self.then_expected_months_are(expected_months)
+
+    @parameterized.expand([
+        'year',
+        'month',
+        'week',
+        'day',
+        'hour',
+        'minute',
+        'second',
+    ])
+    def test_should_reject_easily_mistaken_dateutil_arguments(self, invalid_period):
+        self.given_period_start(datetime(2014, 6, 15))
+        self.given_period_end(datetime(2014, 6, 25))
+        self.given_period_size(**{invalid_period: 1})
+        self.when_period_is_parsed()
+        self.then_period_was_rejected(invalid_period)
+
+    def given_period_start(self, begin):
+        self.period_start = begin
 
-    def test_should_render_10_days_range(self):
-        begin = datetime(2014, 6, 15)
-        end = datetime(2014, 6, 25)
-        dates = list(date.date_range(begin, end))
-        self.assertEquals(10, len(dates))
-        self.assertEquals(begin, dates[0])
-        self.assertEquals(end - timedelta(days=1), dates[-1])
+    def given_period_end(self, end):
+        self.period_end = end
 
-    def test_should_one_date_for_each_month(self):
-        fixtures = [
-            (datetime(2014, 4, 15), datetime(2014, 6, 25),
-             [(2014, 4), (2014, 5), (2014, 6)]),
+    def given_period_size(self, **params):
+        self.period_size = params
 
-            (datetime(2014, 4, 25), datetime(2014, 5, 5),
-             [(2014, 4), (2014, 5)]),
+    def when_month_range_generated(self):
+        self.generated_months = list(date.date_range(self.period_start, self.period_end, **self.period_size))
 
-            (datetime(2014, 4, 5), datetime(2014, 4, 25),
-             [(2014, 4)]),
+    def when_date_range_generated(self):
```

Comment:
Scratch that. We can use `when_period_is_parsed` method instead of both, if we use `list` there, instead of `.next()`